### PR TITLE
Fixes #12303 - Added a correct url to ptables 'Useful template..' link

### PR DESCRIPTION
**Pull request by alongoldboim**
The URL for ptables and provisioning templates was the same.

### CODE REVIEW

File: app/views/templates/_form.html.erb [hunk 1]
```diff
@@ -37,7 +37,8 @@
       <% else -%>
         <%= alert :class => 'alert-success', :header => '',
                   :text  => icon_text("info-sign", (_('Note: %s ') % link_to(_('Useful template functions and macros'),
-                                                                             "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html#4.4.3ProvisioningTemplates", :rel => 'external')).html_safe) %>
+                                                                             type == 'ptable' ? "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html#4.4.4PartitionTables" : "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html#4.4.3ProvisioningTemplates",
```
Comment (shlomizadok) [inline app/views/templates/_form.html.erb]: Seems rather long line. Maybe shorten to something like:

``` ruby
type == 'ptable' ? 
"http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html#4.4.4PartitionTables" : 
"http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html#4.4.3ProvisioningTemplates",
```

Comment (shlomizadok) [inline app/views/templates/_form.html.erb]: Also, there are probably 5-6 places with the url `"http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html` maybe put it in a helper so all places can use it (it may be another PR, though)

Comment (alongoldboim) [inline app/views/templates/_form.html.erb]: Your right @shlomizadok, I think a CONST with http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html should also do the trick. ill add it.

**Comment by alongoldboim** [comment]
Added a helper for foreman documentation url, applied in needed files.

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -409,6 +409,10 @@ def edit_inline(object, property, options = {})
     end
   end
 
+  def foreman_doc_url(text = "")
```
Comment (domcleal) [inline app/helpers/application_helper.rb]: There's already `documentation_button`, I'd suggest this shouldn't have a "foreman_" prefix to match and should be unified.

File: app/views/environments/welcome.html.erb [hunk 3]
```diff
@@ -6,7 +6,7 @@
   <p>
     <%= _('If you are planning to use Foreman as an external node classifier you should provide information about one or more environments.') %>
     <%= _('This information is commonly imported from a pre-existing puppet configuration by the use of the') %>
-    <a href="http://www.theforeman.org/manuals/<%= SETTINGS[:version].short %>/index.html#4.2.2Classes" rel="external"><%= _('Puppet classes and environment importer') %></a>
+    <a href=foreman_doc_url(index.html#4.2.2Classes)" rel="external"><%= _('Puppet classes and environment importer') %></a>
```
Comment (shlomizadok) [inline app/views/environments/welcome.html.erb]: Either switch to `link_to` or use erb here, as this will output as `<a href="foreman_doc_url...`

File: app/helpers/application_helper.rb (hunk 2 above)
Comment (alongoldboim) [inline app/helpers/application_helper.rb]: I'll change the name np, but about unifying I think it would be better to have both of them, one is for a button and one is for url, maybe ill just change the name of this method to documentation_url will that do?

Comment (domcleal) [inline app/helpers/application_helper.rb]: Should the button call this then?

Comment (alongoldboim) [inline app/helpers/application_helper.rb]: Yea that makes sense, ill change the documentation_button.

File: app/helpers/application_helper.rb [hunk 4]
```diff
@@ -381,13 +381,8 @@ def show_parent?(obj)
     (obj.new_record? && obj.class.count > 0) || (!obj.new_record? && obj.class.count > 1)
   end
 
-  def documentation_button(section = nil)
-    url = if section
-            "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html##{section}"
-          else
-            "http://www.theforeman.org/documentation.html##{SETTINGS[:version].short}"
```
Comment (domcleal) [inline app/helpers/application_helper.rb:388]: [fixed] We've lost this conditional, which is used on the dashboard.  I'd say actually replace the link to the manual on the about page with the same link to the main documentation index.. it's a better entry point.

File: app/views/common/_searchbar.html.erb [hunk 5]
```diff
@@ -17,7 +17,7 @@
         <li><%= link_to_function _('Bookmark this search'), "$('#bookmarks-modal').modal();",
                                  {:id => "bookmark", :"data-url"=> main_app.new_bookmark_path(:kontroller => auto_complete_controller_name)} %></li>
         <li><%= link_to(icon_text('question-sign', _('Documentation'), :class => 'icon-black'),
-            "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html#4.1.5Searching",
+                        documentation_url("4.1.5Searching"),
             :rel => 'external', :target => '_blank')
```
Comment (domcleal) [inline app/views/common/_searchbar.html.erb]: Indentation here is different on line 20 versus 21.

File: app/views/environments/welcome.html.erb [hunk 6]
```diff
@@ -6,7 +6,8 @@
   <p>
     <%= _('If you are planning to use Foreman as an external node classifier you should provide information about one or more environments.') %>
     <%= _('This information is commonly imported from a pre-existing puppet configuration by the use of the') %>
-    <a href="http://www.theforeman.org/manuals/<%= SETTINGS[:version].short %>/index.html#4.2.2Classes" rel="external"><%= _('Puppet classes and environment importer') %></a>
+    <%= link_to  _('Puppet classes and environment importer'), documentation_url("4.2.2Classes"), :rel => "external" %>
+
```
Comment (domcleal) [inline app/views/environments/welcome.html.erb]: superfluous line?

File: app/views/templates/_form.html.erb [hunk 7]
```diff
@@ -30,14 +30,19 @@
       <%= render "custom_tabs", :f => f if type == 'ptable' %>
 
       <% if @template.locked? -%>
+        <div id="alert-warning">
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: What's the extra div for?

Comment (alongoldboim) [inline app/views/templates/_form.html.erb]: I added those divs because it makes easier for plugins to make changes on the alerts(defacing).

Comment (domcleal) [inline app/views/templates/_form.html.erb]: Seems unrelated then, can you remove them?

Comment (alongoldboim) [inline app/views/templates/_form.html.erb]: Sure np.
Ill open another PR for those changes.

File: app/helpers/application_helper.rb [hunk 8]
```diff
@@ -409,6 +404,12 @@ def edit_inline(object, property, options = {})
     end
   end
 
+  def documentation_url(section = "")
+    base_url = "http://www.theforeman.org/documentation.html"
```
Comment (domcleal) [inline app/helpers/application_helper.rb]: Missing the version number?

File: app/helpers/application_helper.rb [hunk 9]
```diff
@@ -409,6 +404,12 @@ def edit_inline(object, property, options = {})
...
+    base_url = "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html#" + section unless section.empty?
```
Comment (domcleal) [inline app/helpers/application_helper.rb]: There's no point in setting base_url on line 408 when `section.present?`, so just put it into an if/else block, like the one in the original documentation_button.

File: app/views/templates/_form.html.erb [hunk 10]
```diff
@@ -33,11 +33,11 @@
         <%= alert :class => 'alert-warning', :header => '',
                   :text  => icon_text("warning-sign", (_("Warning: This template is locked. You may only change the associations. Please %s it to customize.") %
                                                             link_to(_('clone'), template_hash_for_member(::Template.find(params[:id]), 'clone_template'))).html_safe) %>
-
       <% else -%>
-        <%= alert :class => 'alert-success', :header => '',
-                  :text  => icon_text("info-sign", (_('Note: %s ') % link_to(_('Useful template functions and macros'),
-                                                                             "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html#4.4.3ProvisioningTemplates", :rel => 'external')).html_safe) %>
+          <%= alert :class => 'alert-success', :header => '',
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: Can you restore the original/correct indentation please?

File: app/views/templates/_form.html.erb [hunk 11]
```diff
@@ -33,11 +33,11 @@
...
+                    :text  => icon_text("info-sign", (_('Note: %s ') % link_to(_('Useful template functions and macros'),
+                                                                               type == 'ptable' ? documentation_url("4.4.4PartitionTables") : documentation_url("4.4.3ProvisioningTemplates"),
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: I don't know if we still want to display this for non-ptable/ProvisioningTemplate models, perhaps @ares has a view on what would be useful to do for plugins?  I'd suggest just leaving out the entire alert block for unknown types.

File: app/views/templates/_form.html.erb [hunk 12]
```diff
@@ -28,16 +28,12 @@
...
-      <% if @template.locked? -%>
-        <%= alert :class => 'alert-warning', :header => '',
-                  :text  => icon_text("warning-sign", (_("Warning: This template is locked. You may only change the associations. Please %s it to customize.") %
-                                                            link_to(_('clone'), template_hash_for_member(::Template.find(params[:id]), 'clone_template'))).html_safe) %>
-
-      <% else -%>
-        <%= alert :class => 'alert-success', :header => '',
-                  :text  => icon_text("info-sign", (_('Note: %s ') % link_to(_('Useful template functions and macros'),
-                                                                             "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html#4.4.3ProvisioningTemplates", :rel => 'external')).html_safe) %>
+      <% if type == 'ptable' || type == 'provisioning_template' %>
```
Comment (dLobatog) [inline app/views/templates/_form.html.erb]: How about rendering this through partials on ptable and provisioning template, like `custom_tabs` in this template?

``` ruby
 <%= render "#{@type_name_plural}/locked_warning" %>
```

**Comment by alongoldboim** [comment]
@dLobatog Moved the alerts to a partial.

**Comment by domcleal** [comment]
I think you've missed the point in @dLobatog's comment, which was that if the `@type_name_plural` was part of the path of the partial that's loaded, you can have a partial per type of template.  Each subclass would have a partial and we'd load the specific one that's needed instead of having conditionals inside one view or partial.

The only catch I'd suggest is that plugins may not have the partial, so we'll either need to merge this into develop only or it needs some error handling if the partial's missing.

**Comment by alongoldboim** [comment]
K now i get what he meant.
Can you give an example for a plugin that creates a new type of template so i could understand the issue better?

**Comment by domcleal** [comment]
https://github.com/theforeman/foreman_remote_execution/blob/master/app/models/job_template.rb

**Comment by alongoldboim** [comment]
@domcleal Thanks,
We could avoid the errors by checking before if a partial exist and if so to render it.
example:
<% if lookup_context.find_all('#{@type_name_plural}/_alerts').any? %>
<%= render '#{@type_name_plural}/_alerts' %>
<% end %>

We would still have one condition added in the view but we would get the functionality of having different partial for every template (we could also add a default if not partial not found) Would something like that be ok?

**Comment by dLobatog** [comment]
@alongoldboim That sounds fine, I think not loading it is fine, maybe we should log something in that case. Can you move that code to a helper, even layout or application helper, then use it there? Something like 

``` ruby
  def render_if_exists(partial)
    partial_exists? = lookup_context.find_all(partial).any? 
    if partial_exists?
      render partial 
    else
      logger.debug("Did not load partial #{partial} because it could not be found.")
    end
  end
```

**Comment by alongoldboim** [comment]
@dLobatog added need partial and helper method.

File: app/helpers/template_paths_helper.rb [hunk 13]
```diff
@@ -23,4 +23,14 @@ def template_hash_for_member(template, member = nil)
...
     send("hash_for_#{member}#{template_route_prefix(template.class).singularize}_path", :id => template)
   end
+
+  def render_if_exists(partial)
+    partial_exists = lookup_context.find_all(partial).any?
+    if partial_exists
+      render :partial => partial.split("_").last
+    else
+      logger.debug("Did not load partial #{partial} because it could not be found.")
+      nil
```
Comment (dLobatog) [inline app/helpers/template_paths_helper.rb]: @alongoldboim I'll remove this nil on merge, it's useless as far as I can see

Comment (dLobatog) [inline app/helpers/template_paths_helper.rb:33]: It looks like this `nil` is useless, can you remove it?

File: app/helpers/application_helper.rb [hunk 14]
```diff
@@ -409,6 +404,15 @@ def edit_inline(object, property, options = {})
     end
   end
 
+  def documentation_url(section = "")
+    if section.empty?
+      base_url = "http://www.theforeman.org/documentation.html##{SETTINGS[:version].short}"
+    else
+      base_url = "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html#" + section
+    end
+    base_url
```
Comment (dLobatog) [inline app/helpers/application_helper.rb]: No need to create a variable, 

``` ruby
def documentation_url(section = '')
   if section.empty?
     "http://www.theforeman.org/documentation.html##{SETTINGS[:version].short}"
   else
     "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html#" + section
  end
end
```

is 100% equivalent without creating any local variable

**Comment by dLobatog** [comment]
@alongoldboim Thanks, see the two quibbles inline, they are really minor. Ready to merge afterwards :smile:

File: app/helpers/template_paths_helper.rb (hunk 13 above)
Comment (alongoldboim) [inline app/helpers/template_paths_helper.rb:33]: The logger.debug method returns you a boolean indicating if it succeeded or not, the render if exists will place true where the alert is if nil isn't placed.

**Comment by alongoldboim** [comment]
@dLobatog Thank you, its ready to merge, added a comment explaining the nil use.

**Comment by dLobatog** [comment]
Merged as 43895066385bcaa8761f8aba532764f04c50952c, thanks @alongoldboim!
